### PR TITLE
Issue #33:  Slack Username Validation

### PR DESCRIPTION
**Pull request by ccostino**
This is the initial changeset with full support and passing tests for the Slack username validation.  There is still one outstanding thing to investigate further with the employee model validation tests as the "allow_values" shoulda matcher method seems to be failing.  I want to look into that a bit further to see if it's something that needs to be reported to the shoulda matchers (and gather full details if so), but otherwise this is ready for review and discussion!  Thank you @jessieay for your help with this and coming up with a cleaner approach, especially with the tests!

### CODE REVIEW

File: app/controllers/employees_controller.rb [hunk 1]
```diff
@@ -6,6 +6,12 @@ def new
   def create
     @employee = Employee.new(employee_params)
 
+    unless EmployeeFinder.new.employee_exists?(@employee.slack_username)
```
Comment (jessieay) [inline app/controllers/employees_controller.rb]: would it be possible to change this to an `if` instead of an `unless`? I find `unless` statements hard to parse because there's a negative - easier to say "if this, then that" rather than "unless not this, then that" (as I write this, I am not sure if that is even correct...more evidence that unless is hard!)

Also, since we are using the same statement on this line and L35, perhaps the logic can be extracted into a private method with a nice name....

File: app/controllers/employees_controller.rb [hunk 2]
```diff
@@ -6,6 +6,12 @@ def new
...
+      flash.now[:error] = "\"#{@employee.slack_username}\" is not a valid employee (that username was not found)."
+      render action: :new
+      return
```
Comment (jessieay) [inline app/controllers/employees_controller.rb]: don't need an explicit return in ruby :)

File: app/controllers/employees_controller.rb [hunk 3]
```diff
@@ -6,6 +6,12 @@ def new
   def create
     @employee = Employee.new(employee_params)
 
+    unless EmployeeFinder.new.employee_exists?(@employee.slack_username)
+      flash.now[:error] = "\"#{@employee.slack_username}\" is not a valid employee (that username was not found)."
```
Comment (jessieay) [inline app/controllers/employees_controller.rb]: how about just `"There is not a slack user with the username \"#{@employee.slack_username}\" in your organization"`

File: spec/models/employee_spec.rb [hunk 4]
```diff
@@ -5,6 +5,11 @@
      subject { create(:employee) }
      it { should validate_presence_of(:slack_username) }
      it { should validate_uniqueness_of(:slack_username) }
+     # These fail - why?
+     #it { should allow_values("test_user_1", "testuser2", "test.user.3").for(:slack_username) }
+     #it { should_not allow_values("TEST USER 1", "test_user_2?", "test'user'3").for(:slack_username) }
```
Comment (jessieay) [inline spec/models/employee_spec.rb]: remove commented code?

File: app/services/employee_finder.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
+require "slack-ruby-client"
+
+class EmployeeFinder
+  def initialize
+    configure_slack
+  end
+
+  def employee_exists?(slack_username)
```
Comment (jessieay) [inline app/services/employee_finder.rb]: so the way we've been writing these services classes, they all use the `run` method to process logic...thoughts on following that pattern here as well?

File: spec/services/slack_user_finder_spec.rb [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+    it "returns false if a user was not found" do
+      fake_slack_user_name = "testusernamethatdoesnotexist"
+
+      user_exists = SlackUserFinder.new(
+        fake_slack_user_name,
+        Slack::Web::Client.new
+      ).user_exists?
+
+      expect(user_exists).to be false
```
Comment (jessieay) [inline spec/services/slack_user_finder_spec.rb]: another way to write this, if we keep the method name as a boolean, would be:


```ruby
user = SlackUserFinder.new(fake_slack_user_name, Slack::Web::Client.new)

expect(user).to be_user_exists
```
one thing I like about doing that is it reveals the quality of boolean method names - now that I wrote it this way, it seems the name of this method is repetitive / off. If we keep it a boolean, perhaps something like `is_in_organization?` would be better...

Source :https://www.relishapp.com/rspec/rspec-expectations/v/2-2/docs/matchers/predicate-matchers#matcher-arguments-are-passed-on-to-the-predicate-method

File: app/controllers/employees_controller.rb (hunk 2 above)
Comment (ccostino) [inline app/controllers/employees_controller.rb]: In this case we do though, without it the controller action was continuing to flow, causing it to add the invalid employee anyway!  Unless (pun not intended, haha!) there's another/better way to immediately stop and return?

File: app/controllers/employees_controller.rb (hunk 3 above)
Comment (ccostino) [inline app/controllers/employees_controller.rb]: More concise, I like it!

File: app/controllers/employees_controller.rb (hunk 2 above)
Comment (jessieay) [inline app/controllers/employees_controller.rb]: we could do an `if` and `elsif` rather than two `if`s!

File: app/services/employee_finder.rb (hunk 5 above)
Comment (ccostino) [inline app/services/employee_finder.rb]: I'm definitely a fan of consistency, but also explicit naming conventions and the purpose of this method isn't to process anything but instead to find something.  If these methods are to be treated as the main entry points for the service classes regardless of what they do, then perhaps a more generic name might be better?

File: spec/models/employee_spec.rb (hunk 4 above)
Comment (ccostino) [inline spec/models/employee_spec.rb]: Yes, once I uncover what's going on with these. :-)

File: app/controllers/employees_controller.rb (hunk 1 above)
Comment (ccostino) [inline app/controllers/employees_controller.rb]: Yes, definitely on both accounts.  I need to ban `unless` from my Ruby vocabulary... :-)

File: spec/services/slack_user_finder_spec.rb (hunk 6 above)
Comment (ccostino) [inline spec/services/slack_user_finder_spec.rb]: Oh interesting, thank you for the heads up with the RSpec info (something I'm still diving into).  As for the name, are you thinking that it should be changed in the `EmployeeFinder` service class?  I'm not sure about `is_in_organization?` only because that sounds like it should accept an argument of an organization to me and we're just checking for the existence of a user in the organization we're a part of when logged in.  I agree that the naming could be improved though, especially in light of how it flows with the Rspec matchers.

File: app/controllers/employees_controller.rb (hunk 2 above)
Comment (ccostino) [inline app/controllers/employees_controller.rb]: Oh, right!

File: spec/models/employee_spec.rb (hunk 4 above)
Comment (ccostino) [inline spec/models/employee_spec.rb]: In thinking about this a bit more, while this would be an interesting investigation to go down I don't think it's necessary at this time for the sake of finishing this PR and moving forward to other things.  If it rears it's head again we can revisit it then.

File: spec/services/employee_finder_spec.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
+require "rails_helper"
+
+describe EmployeeFinder do
+  describe "#employee_exists?" do
+    it "returns true if an employee exists in Slack" do
```
Comment (jessieay) [inline spec/services/employee_finder_spec.rb]: how about `it "returns true if an employee exists in the Slack organization associated with the auth token" do`

it's kinda wordy, but will be helpful for future people reading these tests!

**Comment by jessieay** [comment]
One comment on a spec description - otherwise looks SO GREAT please please merge this is awesome

File: spec/services/employee_finder_spec.rb (hunk 7 above)
Comment (ccostino) [inline spec/services/employee_finder_spec.rb]: Oh definitely - much better to be explicit rather than implicit so I agree!  Consider it done. :-)

**Comment by ccostino** [comment]
Great, thank you so much for looking things over and providing feedback @jessieay, I appreciate it and they've been very helpful!

**Comment by ccostino** [comment]
Branch is now merged!

**Comment by jessieay** [comment]
nice! for future: if you rebase your branch and force push before merging to master, the PR will automatically get closed when you merge (because the commit hashes will line up) :sparkles:
